internal/pkg/deploy/sns.go

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package deploy holds the structures to deploy infrastructure resources.
+// This file defines workload deployment resources.
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+)
+
+var errInvalidTopicARN = errors.New("ARN is not a Copilot SNS topic")
+var errInvalidARN = errors.New("invalid ARN format")
+var errInvalidARNService = errors.New("ARN is not an SNS topic")
+var errInvalidComponent = errors.New("Topic must be initialized with valid app, env, and workload")
+
+// Topic holds information about a Copilot SNS topic and its ARN, ID, and Name.
+type Topic struct {
+	awsARN arn.ARN
+	app    string
+	env    string
+	wkld   string
+
+	name string
+}
+
+func NewTopic(inputARN string, app, env, wkld string) (*Topic, error) {
+	t := &Topic{
+		app:  app,
+		env:  env,
+		wkld: wkld,
+	}
+	parsedARN, err := arn.Parse(inputARN)
+	if err != nil {
+		return nil, errInvalidARN
+	}
+	t.awsARN = parsedARN
+	if err := t.validateAndExtractName(); err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// Name returns the name of the given SNS topic, stripped of its "app-env-wkld" prefix.
+func (t Topic) Name() string {
+	return t.name
+}
+
+// ID returns the resource ID of the topic (the last element of the ARN).
+func (t Topic) ID() string {
+	return t.awsARN.Resource
+}
+
+func (t Topic) ARN() string {
+	return t.awsARN.String()
+}
+
+func (t Topic) Workload() string {
+	return t.wkld
+}
+
+// validateAndExtractName determines whether the given ARN is a Copilot-valid SNS topic ARN and
+// returns the parsed components of the ARN.
+func (t *Topic) validateAndExtractName() error {
```

Comment:
I am a little wary about ignoring these errors...Some of the validations like `Should not include subscriptions to SNS topics.` since it might happen when the users are using copilot supposedly. For the others I wonder if we should still catch and alert the users. Otherwise they might end up finding their topics created by Copilot are missing with no error logs.

Reply:
refactored to make more explicit, see what you think now